commitrmnocb/observation.go

```diff
@@ -0,0 +1,170 @@
+func (p *Plugin) ObserveOffRampNextSeqNums(ctx context.Context) []plugintypes.SeqNumChain {
+	supportsDestChain, err := p.supportsDestChain(p.nodeID)
+	if err != nil {
+		p.lggr.Warnw("call to SupportsDestChain failed", "err", err)
+		return nil
+	}
+
+	if supportsDestChain {
+		sourceChains := p.knownSourceChainsSlice()
+		offRampNextSeqNums, err := p.ccipReader.NextSeqNum(ctx, sourceChains)
+		if err != nil {
+			p.lggr.Warnw("call to NextSeqNum failed", "err", err)
+			return nil
+		}
+
+		if len(offRampNextSeqNums) != len(sourceChains) {
+			p.lggr.Warnf("call to NextSeqNum returned unexpected number of seq nums, got %d, expected %d",
+				len(offRampNextSeqNums), len(sourceChains))
+			return nil
+		}
+
+		result := make([]plugintypes.SeqNumChain, len(sourceChains))
+		for i := range sourceChains {
+			result[i] = plugintypes.SeqNumChain{ChainSel: sourceChains[i], SeqNum: offRampNextSeqNums[i]}
+		}
+
+		return result
+	}
+
+	return nil
+}
+
+// ObserveMerkleRoots Compute the merkle roots for the given sequence number ranges
+func (p *Plugin) ObserveMerkleRoots(ctx context.Context, ranges []ChainRange) []cciptypes.MerkleRootChain {
+	roots := make([]cciptypes.MerkleRootChain, 0)
+	supportedChains, err := p.supportedChains(p.nodeID)
+	if err != nil {
+		p.lggr.Warnw("call to supportedChains failed", "err", err)
+		return nil
+	}
+
+	for _, chainRange := range ranges {
+		if supportedChains.Contains(chainRange.ChainSel) {
+			msgs, err := p.ccipReader.MsgsBetweenSeqNums(ctx, chainRange.ChainSel, chainRange.SeqNumRange)
+			if err != nil {
+				p.lggr.Warnw("call to MsgsBetweenSeqNums failed", "err", err)
+			} else {
+				root, err := p.computeMerkleRoot(ctx, msgs)
+				if err != nil {
+					p.lggr.Warnw("call to computeMerkleRoot failed", "err", err)
+				} else {
+					merkleRoot := cciptypes.MerkleRootChain{
+						ChainSel:     chainRange.ChainSel,
+						SeqNumsRange: chainRange.SeqNumRange,
+						MerkleRoot:   root,
+					}
+					roots = append(roots, merkleRoot)
+				}
+			}
+		}
+	}
+
+	return roots
+}
+
+// computeMerkleRoot Compute the merkle root of a list of messages
+func (p *Plugin) computeMerkleRoot(ctx context.Context, msgs []cciptypes.Message) (cciptypes.Bytes32, error) {
+	hashes := make([][32]byte, 0)
+	sort.Slice(msgs, func(i, j int) bool { return msgs[i].Header.SequenceNumber < msgs[j].Header.SequenceNumber })
+
+	for i, msg := range msgs {
+		// Assert there are no sequence number gaps in msgs
+		if i > 0 {
+			if msg.Header.SequenceNumber != msgs[i-1].Header.SequenceNumber+1 {
+				return [32]byte{}, fmt.Errorf("found non-consecutive sequence numbers when computing merkle root, "+
+					"gap between sequence nums %d and %d, messages: %v", msgs[i-1].Header.SequenceNumber,
+					msg.Header.SequenceNumber, msgs)
+			}
+		}
+
+		msgHash, err := p.msgHasher.Hash(ctx, msg)
+		if err != nil {
+			p.lggr.Warnw("failed to hash message", "msg", msg, "err", err)
+			return cciptypes.Bytes32{}, err
```

Comment:
Lets wrap errors where added context is useful, e.g in this case

```suggestion
			return cciptypes.Bytes32{}, fmt.Errorf("failed to hash message with id %s: %w", msg.ID.String(), err)
```